packages/contracts/test/UnirepApp.test.ts

```diff
@@ -55,12 +61,86 @@ describe('Unirep App', function () {
         await app.deployed()
     })
 
-    it('user sign up', async () => {
+    it('init user status', async () => {
+        expect(await app.initUserStatus(hashUserId))
+            .to.emit(app, 'UserInitSuccess')
+            .withArgs(hashUserId)
+    })
+
+    it('user sign up after init', async () => {
+        const userState = await genUserState(id, app)
+
+        // get user status, should be RegisterStatus.INIT
+        expect(await app.queryUserStatus(hashUserId)).to.equal(1)
+
+        const { publicSignals, proof } = await userState.genUserSignUpProof()
+        expect(await app.userSignUp(publicSignals, proof, hashUserId, false))
+            .to.emit(app, 'UserSignUpSuccess')
+            .withArgs(hashUserId)
+
+        // get user status, should be RegisterStatus.REGISTERED_SERVER
+        expect(await app.queryUserStatus(hashUserId)).to.equal(2)
+
+        userState.stop()
+    })
+
+    it('revert when user sign up multiple times', async () => {
+        const userState = await genUserState(id, app)
+
+        const { publicSignals, proof } = await userState.genUserSignUpProof()
+        expect(
+            app.userSignUp(publicSignals, proof, hashUserId, false)
+        ).to.be.revertedWithCustomError(app, 'UserAlreadySignedUp')
+
+        userState.stop()
+    })
+
+    it('revert when user sign up without init', async () => {
+        // create another random user id
+        const id = new Identity()
+        const hash = crypto.createHash('sha3-224')
+        const hashUserId = `0x${hash
```

Comment:
We could put this in future refactor to modularize these functionalities to a random userId or hashUserId utils helpers.